owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImagePagerAdapter.kt

```diff
@@ -0,0 +1,173 @@
+
+    private fun updateFile(position: Int, file: OCFile) {
+        val fragmentToUpdate = mCachedFragments[position]
+        if (fragmentToUpdate != null) {
+            mObsoleteFragments.add(fragmentToUpdate)
+        }
+        mObsoletePositions.add(position)
+        mImageFiles[position] = file
+    }
+
+    private fun updateWithDownloadError(position: Int) {
+        val fragmentToUpdate = mCachedFragments[position]
+        if (fragmentToUpdate != null) {
+            mObsoleteFragments.add(fragmentToUpdate)
+        }
+        mDownloadErrors.add(position)
+    }
+
+    fun onTransferServiceConnected() {
+        for (fragmentToUpdate in mCachedFragments.values) {
+            fragmentToUpdate.onTransferServiceConnected()
+        }
+    }
+
+    fun clearErrorAt(position: Int) {
+        val fragmentToUpdate = mCachedFragments[position]
+        if (fragmentToUpdate != null) {
+            mObsoleteFragments.add(fragmentToUpdate)
+        }
+        mDownloadErrors.remove(position)
+    }
+
+    override fun getItemPosition(`object`: Any): Int {
+        if (mObsoleteFragments.contains(`object`)) {
+            mObsoleteFragments.remove(`object`)
+            return POSITION_NONE
+        }
+        return super.getItemPosition(`object`)
+    }
+
+    override fun instantiateItem(container: ViewGroup, position: Int): Any {
+        val fragment = super.instantiateItem(container, position)
+        mCachedFragments[position] = fragment as FileFragment
+        return fragment
+    }
+
+    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
+        mCachedFragments.remove(position)
+        super.destroyItem(container, position, `object`)
+    }
+
+    fun pendingErrorAt(position: Int) = mDownloadErrors.contains(position)
+
+    /**
+     * Reset the image zoom to default value for each CachedFragments
+     */
+    fun resetZoom() {
+        val entries: Iterator<FileFragment> = mCachedFragments.values.iterator()
+        while (entries.hasNext()) {
+            val fileFragment = entries.next()
+            if (fileFragment is PreviewImageFragment) {
+                fileFragment.getImageView().setScale(1f, true)
+            }
+        }
+    }
+
+    fun onDownloadEvent(file: OCFile, action: String, success: Boolean) {
+        val position = getFilePosition(file)
+        if (position >= 0) {
+            if (success) {
+                updateFile(position, file)
+            } else {
+                updateWithDownloadError(position)
+            }
+            val fragment = mCachedFragments[position]
+            if (fragment is FileDownloadFragment && success) {
+                // trigger the creation of new PreviewImageFragment to replace current FileDownloadFragment
+                // only if the download succeded. If not trigger an error
+                notifyDataSetChanged()
+            } else fragment?.onSyncEvent(action, success, null)
+        }
+    }
+}
```

Comment:
Not sure if here we should add a final blank line, I didn't see it in many other classes from this PR, but I tell you just in case

Reply:
Yes, we should add a final blank line 👍